src/carousel/index.ts

```diff
@@ -0,0 +1,234 @@
+    this.state.items = this.cloneItems();
+  }
+
+  onUnMount() {
+    utils.WINDOW().removeEventListener('resize', this.updateWindow);
+  }
+
+  updateWindow = () => {
+    this.update();
+  }
+
+  moveNext = () => {
+    this.slideHandler(this.state.currentSlide + this.state.settings.slidesToScroll);
+  }
+
+  movePrevious = () => {
+    this.slideHandler(this.state.currentSlide - this.state.settings.slidesToScroll);
+  }
+
+  cloneItems = () => {
+    const slidesToShow = this.state.settings.slidesToShow;
+    const numCloned = slidesToShow * 2 - 1;
+    const length = this.props.items.length;
+
+    const prior = this.props.items
+      .slice(-numCloned)
+      .map((d, i) => ({ ...d, 'data-index': i - length }));
+    const posterior = this.props.items
+      .slice(0, numCloned)
+      .map((d, i) => ({ ...d, 'data-index': i }));
+    this.props.items.forEach((d, i) => (d['data-index'] = i));
+
+    return prior.concat(this.props.items).concat(posterior);
+  }
+
+  slideHandler = (slide: number) => {
+    const slidesToShow = this.state.settings.slidesToShow;
+    const from = this.state.currentSlide;
+    const length = this.props.items.length;
+    const onEdge = slide >= length || slide <= 0;
+
+    // don't need to consider on edge if speed is 0;
+    if (!this.state.settings.speed) {
+      this.resetCurrentSlideNum(from, slide, length);
+      return;
+    }
+
+    if (!(this.state.transitioning && onEdge)) {
+      // make the transition
+      this.set({ currentSlide: slide, transitioning: true });
+
+      if (onEdge) {
+        // reset to non-cloned slide
+        // tslint:disable-next-line:max-line-length
+        this.animationEndCallback = setTimeout(() => this.resetToRealSlide(from, slide, length), this.state.settings.speed);
+      } else {
+
+        this.refs.track.addEventListener('transitionend', this.disableTransition, false);
+      }
+    }
+  }
+
+  resetToRealSlide = (from: number, to: number, length: number) => {
+    this.state.transitioning = false;
+    this.resetCurrentSlideNum(from, to, length);
+    this.animationEndCallback = null;
+  }
+
+  resetCurrentSlideNum = (from: number, to: number, length: number) => {
+    let currentSlide;
+    if (from < to) {
+      currentSlide = to % length;
+    } else {
+      currentSlide = (to % length + length) % length;
+    }
+    this.set({ currentSlide });
+  }
+
+  disableTransition = () => {
+    this.refs.track.removeEventListener('transitionend', this.disableTransition);
+    this.set({ transitioning: false });
+  }
+
+  trackStyle = () => {
+    let transitionStyles = {};
+    if (this.state.transitioning && this.state.settings.speed) {
+      const speed = this.state.settings.speed;
+      const transition = speed + 'ms ' + 'ease';
+      transitionStyles = {
+        '-webkit-transition': transition,
+        transition,
+        '-ms-transition': transition
+      };
+    }
+
+    return { ...this.getStaticTrackStyle(), ...transitionStyles };
```

Comment:
I think two return statements could work better here, and not declaring a `transistionStyles` variable at all,
and then on this return just doing `return this.getStaticTrackStyle();`